File: CPP/Lesson_209_hand_of_straights.cpp
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 209 -- Hand of Straights
// Category   : Greedy
// Difficulty : Medium
// Study Plan : Day 105
// =============================================================
//
// QUESTION:
//   Can hand be rearranged into groups of size W of consecutive cards?
// =============================================================
#include <vector>
#include <string>
#include <iostream>
#include <stack>
#include <queue>
#include <unordered_map>
#include <unordered_set>
#include <map>
#include <set>
#include <algorithm>
#include <climits>
#include <numeric>
#include <functional>
#include <cmath>
using namespace std;
bool isNStraightHand(vector<int> h,int W){if((int)h.size()%W)return false;map<int,int> c;for(int x:h)c[x]++;for(auto& p:c){if(p.second>0){int cnt=p.second;for(int k=0;k<W;k++){if(c[p.first+k]<cnt)return false;c[p.first+k]-=cnt;}}}return true;}
int main(){cout<<boolalpha<<isNStraightHand({1,2,3,6,2,3,4,7,8},3)<<"\n"<<isNStraightHand({1,2,3,4,5},4)<<"\n";}

File: Python/Lesson_209_hand_of_straights.py
# =============================================================
# MIT License | @analyticswithharry2026
# GitHub  : https://github.com/analyticswithharry
# YouTube : Analytics with Harry
# =============================================================
# Lesson     : 209 -- Hand of Straights
# Category   : Greedy
# Difficulty : Medium
# Study Plan : Day 105
# =============================================================
#
# QUESTION:
#   Can hand be rearranged into groups of size W of consecutive cards?
# =============================================================
from collections import Counter
def isNStraightHand(h,W):
    if len(h)%W: return False
    c=Counter(h)
    for x in sorted(c):
        if c[x]>0:
            cnt=c[x]
            for k in range(W):
                if c[x+k]<cnt: return False
                c[x+k]-=cnt
    return True

if __name__=="__main__":
    print(isNStraightHand([1,2,3,6,2,3,4,7,8],3))
    print(isNStraightHand([1,2,3,4,5],4))
